### ISSUE TITLE
Error after upgrading to keyring 8: "No recommended password was available"

### ISSUE THREAD

**Issue by hjacobs** [jaraco/keyring#196]
Keyring 8 is broken for me (upgraded from PyPI today, keyring 7 worked before). I'm using Xubuntu 15.10 (Xfce).

We just call the usual `keyring.get_password('mai', ring_user)` ("mai" is our app).

Stacktrace:

```
$ mai
Traceback (most recent call last):
  File "/usr/local/bin/mai", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.4/dist-packages/mai/cli.py", line 305, in main
    cli()
  File "/usr/local/lib/python3.4/dist-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.4/dist-packages/click/core.py", line 1037, in invoke
    return Command.invoke(self, ctx)
  File "/usr/local/lib/python3.4/dist-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.4/dist-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/mai/cli.py", line 55, in cli
    login_with_profile(ctx.obj, profile, data.get(profile), awsprofile)
  File "/usr/local/lib/python3.4/dist-packages/mai/cli.py", line 231, in login_with_profile
    saml_xml, roles = saml_login(user, url)
  File "/usr/local/lib/python3.4/dist-packages/mai/cli.py", line 201, in saml_login
    saml_password = keyring.get_password('mai', ring_user)
  File "/usr/local/lib/python3.4/dist-packages/keyring/core.py", line 42, in get_password
    return _keyring_backend.get_password(service_name, username)
  File "/usr/local/lib/python3.4/dist-packages/keyring/backends/fail.py", line 18, in get_password
    raise RuntimeError("No recommended password was available")
RuntimeError: No recommended password was available
```

**Comment by hjacobs** [comment]
Simple way to reproduce:

```
$ python3 -c "import keyring; keyring.get_password('test', 'jdoe')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.4/dist-packages/keyring/core.py", line 42, in get_password
    return _keyring_backend.get_password(service_name, username)
  File "/usr/local/lib/python3.4/dist-packages/keyring/backends/fail.py", line 18, in get_password
    raise RuntimeError("No recommended password was available")
RuntimeError: No recommended password was available
$ python3 --version
Python 3.4.3+
```

**Comment by hjacobs** [comment]
Setting the password also fails, i.e. it appears to not support the fallback "file" backend anymore (but it should!):

```
$ python3 -c "import keyring; keyring.set_password('test', 'jdoe', 'mypass')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.4/dist-packages/keyring/core.py", line 48, in set_password
    _keyring_backend.set_password(service_name, username, password)
  File "/usr/local/lib/python3.4/dist-packages/keyring/backends/fail.py", line 18, in get_password
    raise RuntimeError("No recommended password was available")
RuntimeError: No recommended password was available
```

**Comment by hjacobs** [comment]
Apparently the "file" backend was deleted in this commit https://github.com/jaraco/keyring/commit/1e422ed5db99ea2d3020de8ff8c7a3cd7c1f13d8 . Could you please document how to use the "file" backend now?

A "deprecation" warning would be really nice for such breaking changes in the future.

**Comment by hjacobs** [comment]
Installing "keyrings.alt" solves the problem:

```
$ sudo pip3 install keyrings.alt
Collecting keyrings.alt
  Downloading keyrings.alt-1.1-py2.py3-none-any.whl
Installing collected packages: keyrings.alt
Successfully installed keyrings.alt-1.1
You are using pip version 7.1.2, however version 8.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
~/workspace/aws-saml-login (master) $ python3 -c "import keyring; keyring.set_password('test', 'jdoe', 'mypass')"
~/workspace/aws-saml-login (master) $ python3 -c "import keyring; print(keyring.get_password('test', 'jdoe'))"
mypass
```

Please document this behavior as using the "keyring" package without "keyrings.alt" will break on all platforms without a running key agent.

I would really like to see the "file" backend as part of the core "keyring" package again so it would be save to use on all platforms (otherwise every app/script needs a try/except around your "keyring" package) :disappointed:

**Comment by hjacobs** [comment]
Please also consider changing the error message to "No recommended password BACKEND was available" (as the backend is missing, not the password itself).

**Comment by jaraco** [comment]
Sorry for the confusion and trouble.

The change is prominently documented in that the bump from 7.x to 8.0 signals a backward-incompatible change. Looking at [the changelog](https://pythonhosted.org/keyring/history.html), it clearly indicates what the change is and what actions are necessary to maintain compatibility.

I did consider adding a deprecation warning, but I realized that the transition would be difficult to manage with forward compatibility. Since the workaround is as simple as installing an additional package, I opted to take the expedient route.

**Comment by jaraco** [comment]
Released as 8.0.1.

**Comment by hjacobs** [comment]
Can you elaborate why the encrypted file backend is considered "not preferred" (the changelog does not tell me the "why")? It's the only backend which reliably works on hosts without a GUI. The security should also be "OK" for single user systems (multi user systems/servers have a ton of different issues regarding personal credentials anyway).

I can totally understand moving exotic backends and the plaintext backend into another package, but I believe that this Python module should run "out of the box" on systems e.g. without a X server.

Basically I would now always need to distribute my command line tools with two dependencies ("keyring" and "keyrings.alt")...

**Comment by jaraco** [comment]
There are a couple of reasons.
- There are known but unresolved deficiencies with the CryptedFile implementation (see #66).
- The CryptedFile keyring requires user interaction, which means the user experience is different than on other platforms.

The former is more concerning to me than the latter. I'd be much more inclined to accept another keyring backend for Linux non-GUI environments that had some of the same characteristics as the Windows and OS X backends (security of keys managed by something other than keyring and unlocked through the user's authentication process), though I doubt that such a thing exists.

I seem to recall someone saying that the SecretService backend doesn't rely on a GUI implementation. Perhaps there's a configuration whereby you could store your passwords using that backend instead.

**Comment by hjacobs** [comment]
Thanks for the explanation, I'll look into it...

**Comment by mitya57** [comment]
The SecretService backend can work without GUI integration, if the default collection is unlocked on login (i.e. via a PAM module).

**Comment by jaraco** [comment]
@mitya57 Would you be willing to put together a mini-HOWTO in the readme for Linux cli users?

**Comment by mitya57** [comment]
Sure, probably not today though.

**Comment by jaraco** [comment]
:thumbsup:
